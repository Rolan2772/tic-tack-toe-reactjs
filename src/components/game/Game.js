import React from "react";
import Board from "../board/Board";
import utils from "../../utils/utils";
import Button from '@material-ui/core/Button'
import Typography from '@material-ui/core/Typography';
import Grid from "@material-ui/core/Grid";
import {withStyles} from '@material-ui/core/styles';

const styles = () => ({
    historyControl: {
        margin: '2px'
    }
});

export const GameStatus = props => {
    return (
        <Typography variant="h6">{props.text}</Typography>
    );
};

export const GameHistory = props => {

    const renderMove = move => {
        return (
            <Button
                key={move}
                // className={props.classes.historyControl}
                variant="contained"
                onClick={() => this.jumpTo(move)}>
                {move > 0 ? `Go to move # ${move}` : `Go to game start`}
            </Button>
        )
    };

    const renderMoves = (moves) => {
        return moves.map((step, move) => renderMove(move));
    };

    return (
        <React.Fragment>
            <Grid item xs={3} md={4} lg={5}>
            </Grid>
            <Grid item xs={6} md={4} lg={2}>
                <Grid container direction={"column"}>
                    {renderMoves(props.history)}
                </Grid>
            </Grid>
            <Grid item xs={3} md={4} lg={5}>
            </Grid>
        </React.Fragment>
    );
};


export class Game extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            history: [{
                squares: Array(9).fill(null),
            }],
            step: 0,
            isX: true,
            winner: null
        }
    }

    jumpTo(step) {
        this.setState({
            step: step,
            isX: (step % 2) === 0,
        });
    }

    handleClick(i) {
        const history = this.state.history;
        const current = history[this.state.step];
        const squares = current.squares.slice();
        if (squares[i] || this.state.winner) {
            return;
        }
        squares[i] = this.state.isX ? 'X' : 'O';
        history.push({squares: squares});
        this.setState({
            history: history,
            step: history.length - 1,
            isX: !this.state.isX,
            winner: utils.findWinner(squares)
        });
    }

    render() {
        const {history, step} = this.state;
        const currentState = history[step];

        const status = this.state.winner
            ? `Winner: ${this.state.winner}`
            : `Next move: ${this.state.isX ? 'X' : 'O'}`;

        return (
            <Grid container direction="column">
                <Grid container justify="center">
                    <GameStatus text={status}/>
                </Grid>
                <Grid>
                    <Board squares={currentState.squares} onClick={(i) => this.handleClick(i)}/>
                </Grid>
                <Grid container>
                    <GameHistory history={history}/>
                </Grid>
            </Grid>
        );
    }
}

export default withStyles(styles)(Game);
