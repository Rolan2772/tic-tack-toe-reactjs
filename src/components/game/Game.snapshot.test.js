import React from 'react';
import Game from './Game';
import renderer from 'react-test-renderer';

describe('<Game />', () => {

    test('should match snapshot', () => {
        const component = renderer.create(<Game/>);

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

