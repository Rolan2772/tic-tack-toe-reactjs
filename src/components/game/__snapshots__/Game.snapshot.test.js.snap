// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> should match snapshot 1`] = `
<div
  className="MuiGrid-container-2 MuiGrid-direction-xs-column-5"
>
  <div
    className="MuiGrid-container-2 MuiGrid-justify-xs-center-19"
  >
    <h6
      className="MuiTypography-root-99 MuiTypography-h6-116"
    >
      Next move: X
    </h6>
  </div>
  <div
    className=""
  >
    <div
      className="Board-boardMargin-135"
    >
      <div
        className="MuiGrid-container-2 MuiGrid-justify-xs-center-19"
      >
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
      </div>
      <div
        className="MuiGrid-container-2 MuiGrid-justify-xs-center-19"
      >
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
      </div>
      <div
        className="MuiGrid-container-2 MuiGrid-justify-xs-center-19"
      >
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-outlined-145 Square-cell-136"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-container-2"
  >
    <div
      className="MuiGrid-item-3 MuiGrid-grid-xs-3-33 MuiGrid-grid-md-4-62 MuiGrid-grid-lg-5-77"
    />
    <div
      className="MuiGrid-item-3 MuiGrid-grid-xs-6-36 MuiGrid-grid-md-4-62 MuiGrid-grid-lg-2-74"
    >
      <div
        className="MuiGrid-container-2 MuiGrid-direction-xs-column-5"
      >
        <button
          className="MuiButtonBase-root-163 MuiButton-root-137 MuiButton-contained-148 MuiButton-raised-151"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-138"
          >
            Go to game start
          </span>
          <span
            className="MuiTouchRipple-root-166"
          />
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-item-3 MuiGrid-grid-xs-3-33 MuiGrid-grid-md-4-62 MuiGrid-grid-lg-5-77"
    />
  </div>
</div>
`;
